smt/surrogate_models/krg_based.py

```diff
@@ -400,233 +392,123 @@ def project_values(self, x):
-        vt = self.vartype
-        # continuous or integer variables only (float) => no categorical one
-        if isinstance(xlim[0][0], np.float64):
-            for ty in vt:
-                if not (ty == 0 or ty == 1):
-                    raise Exception("xlimits used an incorrect type")
-        # Not a float (string or list) => must have a categorical variable
-        elif isinstance(xlim[0][0], np.str_) or isinstance(xlim[0], list):
-            rais = 1
-            xlim = np.zeros((np.size(vt), 2))
-            ind_r = 0
-            ind_o = -1
-            tmp = 0
-            count = 0
-            for ty in vt:
-                # if cate : add dimensions
-                if not (ty == 0 or ty == 1):
-                    rais = 0
-                    xlim[ind_r] = [0, 1]
-
-                    # if (cate,n) we should have n labels
-                    if ty == tmp:
-                        count = count + 1
-                    else:
-                        ind_o = ind_o + 1
-                        tmp = ty
-                        count = 0
-                    try:
-                        err = xlimits[ind_o][count]
-                    except:
-                        raise Exception("missing labels in xlimits")
-
+            return np.array(xlimits)
+
+        xlims = []
+        for i, vt in enumerate(self.options["vartype"]):
+            if vt == "cont" or vt == "int":
+                xlims.append(xlimits[i])
+            elif isinstance(vt, tuple) and vt[0] == "cate":
+                if vt[1] == len(xlimits[i]):
+                    xlims.extend(vt[1] * [[0, 1]])
                 else:
-                    # if it is not a categorical variable : recopy bounds
-                    no_cate = 0
-                    while no_cate == 0:
-                        try:
-                            ind_o = ind_o + 1
-                            xlim[ind_r] = xlimits[ind_o]
-                            no_cate = 1
-                        except:
-                            if ind_o == np.size(vt) + 1:
-                                raise Exception("xlimits used an incorrect type")
-                ind_r = ind_r + 1
-            if rais == 1:
-                raise Exception("xlimits used an incorrect type")
-
-        return xlim
+                    raise ValueError(
+                        "Bad xlimits for categorical var[{}] "
+                        "should have {} categories, got only {} in {}".format(
+                            k, vt[1], len(xlimits[i]), xlimits[i]
+                        )
+                    )
+            else:
+                raise ValueError(
+                    "Bad var type specification: "
+                    "should be 'cont', 'int' or ('cate', n), got {}".format(vt)
+                )
+        return np.array(xlims)
 
     def assign_labels(self, x, xlimits):
         """
         This function reduce inputs from relaxed space to original space by 
```

Comment:
This function do not assign labels to categorical variables.
Therefore, it do not reduce inputs from the relaxed space to the original space but to another intermediate space where categorical variables are "coded" with numbers.